crates/communication/src/server/databases/router.rs

```diff
@@ -0,0 +1,358 @@
+use std::collections::{hash_map::Entry, BTreeMap, HashMap};
+
+use tokio::{
+    spawn,
+    sync::mpsc::{Receiver, Sender},
+    task::JoinHandle,
+};
+
+use crate::server::messages::{
+    DatabaseRequest, Path, Response, TextualDatabaseResponse, TextualResponse, Type,
+};
+
+use super::{Client, ClientRequest, Request};
+
+pub fn router(mut request_receiver: Receiver<Request>) -> JoinHandle<()> {
+    spawn(async move {
+        let mut request_channels_of_cyclers = HashMap::new();
+        let mut cached_cycler_instances = HashMap::new();
+
+        while let Some(request) = request_receiver.recv().await {
+            match request {
+                Request::ClientRequest(request) => {
+                    forward_client_request_to_provider(
+                        request,
+                        &request_channels_of_cyclers,
+                        &mut cached_cycler_instances,
+                    )
+                    .await
+                }
+                Request::RegisterCycler {
+                    cycler_instance,
+                    fields,
+                    request_sender,
+                } => {
+                    request_channels_of_cyclers.insert(cycler_instance, (fields, request_sender));
+                }
+            }
+        }
+    })
+}
+
+async fn forward_client_request_to_provider(
+    request: ClientRequest,
+    request_channels_of_cyclers: &HashMap<String, (BTreeMap<Path, Type>, Sender<ClientRequest>)>,
+    cached_cycler_instances: &mut HashMap<(Client, usize), String>,
+) {
+    match &request.request {
+        DatabaseRequest::GetFields { id } => {
+            // TODO: directly generate BTreeMap in SerializeHierarchy
```

Comment:
Yes, this is still relevant. Since it is already mentioned in Ideas for next iteration, I removed the `TODO` from the code. Also the second `TODO` regarding failed unsubscription.